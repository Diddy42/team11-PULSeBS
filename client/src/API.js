async function userLogin(username, password) {
    return new Promise((resolve, reject) => {
        fetch('http://localhost:80/server/api/login.php', {     //Da verificare
            method: 'POST',
            headers: {
                //'Content-Type': 'application/json',
                'Content-Type' : 'application/x-www-form-urlencoded; charset=UTF-8',
            },
            body: JSON.stringify({username: username, password: password}),
        }).then((response) => {
            response.json().then((obj) => { 
                console.log(obj);
                if(obj === 0 ) {        //Da modificare in base a quanto ritorna il server. Sarebbe utile che ritorni l'id o anche nome e cognome la visualizzare, e il ruolo
                    reject(obj);
                } else {
                    resolve(obj);    
                }
            })
            .catch((err) => { reject({ errors: [{ param: "Application", msg: "Cannot parse server response" }] }) });
        }).catch((err) => { reject({ errors: [{ param: "Server", msg: "Cannot communicate" }] }) });
    });
}

async function logout(){
	return new Promise((resolve, reject) => {
        fetch('http://localhost:80/server/api/logout.php')      //Da verificare
        .then((response) => {
			if(response.ok){
				resolve();
			}
			else{
				reject();
			}
       })
       .catch((err) => { 
      		 reject(err); 
       });
    });
}


async function getStudentLectures(studentId) {
    const url = "test"                  //Da definire
    const response = await fetch(url);
    const lectures = await response.json();
    if(response.ok) {
        return lectures;
    } else {
        let err = {status: response.status, errorObj: lectures};
        throw err; 
    }
}

async function getBooking(lectureId) {
    const url = "test"                  //Da definire
    const response = await fetch(url);
    const booking = await response.json();
    if(response.ok) {
        return booking;
    } else {
        let err = {status: response.status, errorObj: booking};
        throw err; 
    }
}

async function bookASeat(lectureId, studentId) {
    const url = "test"                  //Da definire
    const response = await fetch(url, {
        method: 'POST',
        headers: {
            //'Content-Type': 'application/json',
            'Content-Type' : 'application/x-www-form-urlencoded; charset=UTF-8',
        },
        body: JSON.stringify({lectureId: lectureId, studentId: studentId}),
    });
    
    try{
        const resJ = await response.json();
        if(response.ok) {
		    return resJ;
		} else {
		    let err = {status: response.status, errorObj: resJ};
		    throw err; 
		}
    }
    catch(e){
        console.log("Error in booking a seat: " + e);
        return 0;   //meaning: no new customer
    }
}

const API = {
    userLogin, logout, getStudentLectures, getBooking, bookASeat
};
export default API;
