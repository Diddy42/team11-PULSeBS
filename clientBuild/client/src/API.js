async function userLogin(username, password) {
    return new Promise((resolve, reject) => {
        fetch('http://localhost/server/src/api/Login.php', {
            method: 'POST',
            headers: {
                //'Content-Type': 'application/json',
                'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
            },
            body: JSON.stringify({ username: username, password: password }),
        }).then((response) => {
            response.json().then((obj) => {
                /*
                if(obj === 0 ) {        //Da modificare in base a quanto ritorna il server. Sarebbe utile che ritorni l'id o anche nome e cognome la visualizzare, e il ruolo
                    reject(obj);
                } else {
                    resolve(obj);    
                }
                */
                resolve(obj);
            })
                .catch((err) => { reject({ errors: [{ param: "Application", msg: "Cannot parse server response" }] }) });
        }).catch((err) => { reject({ errors: [{ param: "Server", msg: "Cannot communicate" }] }) });
    });
}

async function logout() {
    return new Promise((resolve, reject) => {
        fetch('http://localhost:80/server/api/logout.php')      //Da verificare
            .then((response) => {
                if (response.ok) {
                    resolve(0);
                }
                else {
                    reject();
                }
            })
            .catch((err) => {
                reject(err);
            });
    });
}

//Ritorna le lezioni prenotabili dallo studente
async function getBookableStudentLectures(studentId) {
    const url = "http://localhost/server/bookableLessons/" + studentId;
    const response = await fetch(url);
    const lectures = await response.json();
    if (response.ok) {
        if (lectures === 0) {
            return [];
        }
        return lectures;
    } else {
        let err = { status: response.status, errorObj: lectures };
        throw err;
    }
}

//Ritorna le prenotazioni effettuate dagli studenti per la lezione indicata (teacher side)
async function getBooking(lectureId) {
    const url = "http://localhost/server/bookedStudentsForLecture/" + lectureId;
    const response = await fetch(url);
    const booking = await response.json();
    if (response.ok) {
        if (booking === 0) {
            return [];
        }
        return booking;
    } else {
        let err = { status: response.status, errorObj: booking };
        throw err;
    }
}

// API to retrieve all the bookings of a student
async function getStudentBookings(studentId) {
    const url = "http://localhost/server/studentBookings/" + studentId;
    const response = await fetch(url);
    const booking = await response.json();
    if (response.ok) {
        if (booking === 0) {
            return [];
        }
        return booking;
    } else {
        let err = { status: response.status, errorObj: booking };
        throw err;
    }
}

//API per prenotare un posto a lezione
async function bookASeat(lectureId, studentId, date) {
    const url = "http://localhost/server/insertLecture"
    const response = await fetch(url, {
        method: 'POST',
        headers: {
            //'Content-Type': 'application/json',
            'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
        },
        body: JSON.stringify({
            idLesson: lectureId,
            idUser: studentId,
            date: date
        }),
    });

    try {
        const resJ = await response.json();
        if (response.ok) {
            return resJ;
        } else {
            let err = { status: response.status, errorObj: resJ };
            throw err;
        }
    }
    catch (e) {
        console.log("Error in booking a seat: " + e);
        return 0;   //meaning: no new customer
    }
}

// API to retrieve all the lectures of a teacher
async function getTeacherLectures(teacherId) {
    const url = "http://localhost/server/scheduledLecturesForTeacher/" + teacherId
    const response = await fetch(url);
    const lectures = await response.json();
    if (response.ok) {
        if (lectures === 0) {
            return [];
        }
        return lectures;
    } else {
        let err = { status: response.status, errorObj: lectures };
        throw err;
    }
}

async function deleteBooking(bookingId) {
    const url = "http://localhost/server/updateBooking/" + bookingId;
    return new Promise((resolve, reject) => {
        fetch(url, {
            method: 'PUT',
            headers: {
                //'Content-Type': 'application/json',
                'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
            },
            body: JSON.stringify({
            }),
        }).then((response) => {
            if (response.ok) {
                resolve(0);
            } else {
                reject({ errore: "Error in deleting a booking" });
            }
        }).catch((err) => { reject({ errors: [{ param: "Server", msg: "Cannot communicate" }] }) });
    });
}

async function deleteLecture(lectureId) {
    const url = "http://localhost/server/deleteLecture/" + lectureId;
    const response = await fetch(url, {
        method: 'PUT',
        headers: {
            //'Content-Type': 'application/json',
            'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
        },
        body: JSON.stringify({
        }),
    });
    const result = await response.json();
    if (response.ok) {
        if (result === 1) {
            return 1;
        } else {
            console.log("Too late to delete the lecture");
            return -1;
        }
    } else {
        let err = { status: response.status, errorObj: result };
        throw err;
    }
}

async function changeToOnline(lectureId) {
    const url = "http://localhost/server/changeToOnline/" + lectureId;
    const response = await fetch(url, {
        method: 'PUT',
        headers: {
            //'Content-Type': 'application/json',
            'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
        },
        body: JSON.stringify({
        }),
    });
    const result = await response.json();
    if (response.ok) {
        if (result === 1) {
            return 1;
        } else {
            console.log("Too late to change the lecture");
            return -1;
        }
    } else {
        let err = { status: response.status, errorObj: result };
        throw err;
    }
}

async function getAllCourses() {
    const url = "http://localhost/server/courses";

    const response = await fetch(url);
    const courses = await response.json();
    if (response.ok) {
        if (courses === 0) {
            return [];
        }
        return courses;
    } else {
        let err = { status: response.status, errorObj: courses };
        throw err;
    }
}

const ALL_COURSES_FILTER = "L.idCourse";
async function getBookingStatisticsByMonth(courseIds) {
    const url = `http://localhost/server/bookingStatistics/?filterTime=year,monthOfYear&filterCourse=${courseIds.toString()}`;

    const response = await fetch(url);
    const stats = await response.json();
    if (response.ok) {
        if (stats === 0) {
            return [];
        }
        return stats;
    } else {
        let err = { status: response.status, errorObj: stats };
        throw err;
    }
} 
async function getBookingStatisticsByWeek(courseIds) {
    const url = `http://localhost/server/bookingStatistics/?filterTime=year_month_week&filterCourse=${courseIds.toString()}`;

    const response = await fetch(url);
    const stats = await response.json();
    if (response.ok) {
        if (stats === 0) {
            return [];
        }
        return stats;
    } else {
        let err = { status: response.status, errorObj: stats };
        throw err;
    }
}
async function getBookingStatisticsByLesson(courseIds) {
    const url = `http://localhost/server/bookingStatistics/?filterTime=L.idLesson&filterCourse=${courseIds.toString()}`;

    const response = await fetch(url);
    const stats = await response.json();
    if (response.ok) {
        if (stats === 0) {
            return [];
        }
        return stats;
    } else {
        let err = { status: response.status, errorObj: stats };
        throw err;
    }
}

async function getCoursesOfTeacher(teacherId) {
    const url = "http://localhost/server/teacherCourses/" + teacherId;

    const response = await fetch(url);
    const courses = await response.json();
    if (response.ok) {
        if (courses === 0) {
            return [];
        }
        return courses;
    } else {
        let err = { status: response.status, errorObj: courses };
        throw err;
    }
}

const API = {
    userLogin, logout, getBookableStudentLectures, getBooking, getStudentBookings, bookASeat, deleteBooking, getTeacherLectures, deleteLecture, changeToOnline, getAllCourses, 
    getBookingStatisticsByMonth, getBookingStatisticsByWeek, getBookingStatisticsByLesson, ALL_COURSES_FILTER, getCoursesOfTeacher
};
export default API;
