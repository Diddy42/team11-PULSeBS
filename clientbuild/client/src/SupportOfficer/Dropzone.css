.Dropzone {
  height: 300px;
  width: 300px;
  background-color: #fff;
  border: 2px dashed rgb(187, 186, 186);
  border-radius: 15%;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  font-size: 16px;
}
  
.Highlight {
  background-color: rgb(235, 235, 235);
}

.UploadIcon {
  opacity: 0.4;
  height: 64px;
  width: 64px;
}

.FileIcon {
  color: #3F51B5;
  opacity: 0.8;
  height: 64px;
  width: 64px;
}

.FileInput {
  display: none;
}