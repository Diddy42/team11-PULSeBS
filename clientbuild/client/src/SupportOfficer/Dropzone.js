import React, { Component } from 'react'
import { FaFileUpload, FaFileAlt } from "react-icons/fa";
import './Dropzone.css'

class Dropzone extends Component {

    constructor(props) {
        super(props);
        this.state = {
            highlight: false
        };

        this.fileInputRef = React.createRef();

        this.openFileDialog = this.openFileDialog.bind(this);
        this.onFileAdded = this.onFileAdded.bind(this);
        this.onDragOver = this.onDragOver.bind(this);
        this.onDragLeave = this.onDragLeave.bind(this);
        this.onDrop = this.onDrop.bind(this);
    }

    openFileDialog() {
        if (this.props.disabled) return;
        this.fileInputRef.current.click();
    }

    onFileAdded(event) {
        if (this.props.disabled) return;
        if (event.target.files.length > 0) {
            const file = event.target.files[0];

            if (this.props.onFileAdded) {
                this.props.onFileAdded(file);
            }
        }
    }

    onDragOver(event) {
        event.preventDefault();
        if (this.props.disabled) return;
        this.setState({ hightlight: true });
    }

    onDragLeave(event) {
        this.setState({ hightlight: false });
    }

    onDrop(event) {
        event.preventDefault();
        if (this.props.disabled) return;
        if (event.dataTransfer.files.length > 0) {
            const file = event.dataTransfer.files[0];
            if (file.type === "text/csv") {
                if (this.props.onFileAdded) {
                    this.props.onFileAdded(file);
                }
            }
        }
        this.setState({ hightlight: false });
    }

    render() {
        return <>
            <div
                className={`Dropzone ${this.state.hightlight ? "Highlight" : ""}`}
                onDragOver={this.onDragOver}
                onDragLeave={this.onDragLeave}
                onDrop={this.onDrop}
                onClick={this.openFileDialog}
                style={{ cursor: this.props.disabled ? "default" : "pointer" }}
            >
                <input
                    ref={this.fileInputRef}
                    className="FileInput"
                    type="file"
                    accept=".csv"
                    onChange={this.onFileAdded}
                />
                {
                    this.props.file === undefined ?
                        <FaFileUpload className="UploadIcon my-3" />
                        :
                        <FaFileAlt className="FileIcon my-3" />
                }
                <span className="Filename">{this.props.file !== undefined ? this.props.file.name : "Choose a file"}</span>
            </div>
        </>
    }
}

export default Dropzone;