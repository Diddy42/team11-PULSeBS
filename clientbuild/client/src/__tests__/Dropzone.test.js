import TestRenderer from 'react-test-renderer';
import React from 'react';
import Dropzone from '../SupportOfficer/Dropzone';
import API from '../API';
import { render } from '@testing-library/react';
import { shallow } from 'enzyme';
import { fireEvent } from '@testing-library/react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

test('renders Dropzone page', () => {
    const tree = TestRenderer.create(<Dropzone disabled={false}/>).toJSON();
    expect(tree).toMatchSnapshot();
  });

test("render Dropzone with file", async () => {
    const tree = TestRenderer.create(<Dropzone disabled={false} onFileAdded={jest.fn()} file={{name: "test.csv", type: "text/csv", name: "test"}}/>).toJSON();
    expect(tree).toMatchSnapshot();
});

const onFileAddedMock = jest.fn();
test("onDrop", async () => {
    const component = await shallow(<Dropzone disabled={false} onFileAdded={jest.fn()}/>);
    const instance = await component.instance();
    await instance.onDrop({dataTransfer: {files: [{name: "test.csv", type: "text/csv"}, {type: "text/csv"}]}, preventDefault: jest.fn()});
    expect(instance.props.onFileAdded).toHaveBeenCalledTimes(1);
    expect(instance.state.highlight).toBe(false);
    await instance.onDrop({dataTransfer: {files: [{name: "test.json", type: "text/JSON"}]}, preventDefault: jest.fn()});
    expect(instance.props.onFileAdded).toHaveBeenCalledTimes(1);
    expect(instance.state.highlight).toBe(false);
    await instance.onDrop({dataTransfer: {files: []}, preventDefault: jest.fn()});
    expect(instance.props.onFileAdded).toHaveBeenCalledTimes(1);
    expect(instance.state.highlight).toBe(false);
});

test("onDrop with undefined onFileAdded", async () => {
    const component = await shallow(<Dropzone disabled={false} onFileAdded={undefined}/>);
    const instance = await component.instance();
    await instance.onDrop({dataTransfer: {files: [{name: "test.csv", type: "text/csv"}, {type: "text/csv"}]}, preventDefault: jest.fn()});
    expect(instance.props.onFileAdded).toBe(undefined);
    expect(instance.state.highlight).toBe(false);
});

test("onDrop with disabled true", async () => {
    const component = await shallow(<Dropzone disabled={true} onFileAdded={undefined}/>);
    const instance = await component.instance();
    await instance.onDrop({dataTransfer: {files: [{name: "test.csv", type: "text/csv"}, {type: "text/csv"}]}, preventDefault: jest.fn()});
    expect(instance.props.onFileAdded).toBe(undefined);
    expect(instance.state.highlight).toBe(false);
});

test("onDragLeave", async () => {
    const component = await shallow(<Dropzone disabled={false} onFileAdded={onFileAddedMock}/>);
    const instance = await component.instance();
    await instance.onDragLeave({dataTransfer: {files: [{name: "test.csv", type: "text/csv"}]}, preventDefault: jest.fn()});
    expect(instance.state.highlight).toBe(false);
});

test("onDragOver", async () => {
    const component = await shallow(<Dropzone disabled={false} onFileAdded={onFileAddedMock}/>);
    const instance = await component.instance();
    expect(instance.state.highlight).toBe(false);
    await instance.onDragOver({target: {files: [{name: "test.csv", type: "text/csv"}]}, preventDefault: jest.fn()});
    expect(instance.state.highlight).toBe(true);
});

test("onDragOver disabled true", async () => {
    const component = await shallow(<Dropzone disabled={true} onFileAdded={onFileAddedMock}/>);
    const instance = await component.instance();
    expect(instance.state.highlight).toBe(false);
    await instance.onDragOver({target: {files: [{name: "test.csv", type: "text/csv"}]}, preventDefault: jest.fn()});
    expect(instance.state.highlight).toBe(false);
});

test("onFileAdded", async () => {
    const component = await shallow(<Dropzone disabled={false} onFileAdded={jest.fn()}/>);
    const instance = await component.instance();
    expect(instance.state.highlight).toBe(false);
    await instance.onFileAdded({target: {files: [{name: "test.csv", type: "text/csv"}]}, preventDefault: jest.fn()});
    expect(instance.props.onFileAdded).toHaveBeenCalledTimes(1);
    await instance.onFileAdded({target: {files: []}, preventDefault: jest.fn()});
    expect(instance.props.onFileAdded).toHaveBeenCalledTimes(1);
});

test("onFileAdded undefined onFileAdded", async () => {
    const component = await shallow(<Dropzone disabled={false} onFileAdded={undefined}/>);
    const instance = await component.instance();
    expect(instance.state.highlight).toBe(false);
    await instance.onFileAdded({target: {files: [{name: "test.csv", type: "text/csv"}]}, preventDefault: jest.fn()});
    expect(instance.props.onFileAdded).toBe(undefined);
});

test("onFileAdded disabled", async () => {
    const component = await shallow(<Dropzone disabled={true} onFileAdded={jest.fn()}/>);
    const instance = await component.instance();
    expect(instance.state.highlight).toBe(false);
    await instance.onFileAdded({target: {files: [{name: "test.csv", type: "text/csv"}]}, preventDefault: jest.fn()});
    expect(instance.props.onFileAdded).toHaveBeenCalledTimes(0);
});

test("openFileDialog", async () => {
    const component = await shallow(<Dropzone disabled={false} onFileAdded={true}/>);
    const instance = await component.instance();
    instance.fileInputRef = {current: {click: jest.fn()}};
    await instance.openFileDialog({target: {files: [{name: "test.csv", type: "text/csv"}]}, preventDefault: jest.fn()});
    expect(instance.fileInputRef.current.click).toHaveBeenCalledTimes(1);
});

test("openFileDialog disabled true", async () => {
    const component = await shallow(<Dropzone disabled={true} onFileAdded={true}/>);
    const instance = await component.instance();
    instance.fileInputRef = {current: {click: jest.fn()}};
    await instance.openFileDialog({target: {files: [{name: "test.csv", type: "text/csv"}]}, preventDefault: jest.fn()});
    expect(instance.fileInputRef.current.click).toHaveBeenCalledTimes(0);
});