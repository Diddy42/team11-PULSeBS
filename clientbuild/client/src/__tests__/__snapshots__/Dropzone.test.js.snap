// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Dropzone with file 1`] = `
Array [
  <div
    className="Dropzone "
    onClick={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    <input
      accept=".csv"
      className="FileInput"
      id="uploadCaptureInputFile"
      onChange={[Function]}
      type="file"
    />
    <svg
      className="FileIcon my-3"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 384 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M224 136V0H24C10.7 0 0 10.7 0 24v464c0 13.3 10.7 24 24 24h336c13.3 0 24-10.7 24-24V160H248c-13.2 0-24-10.8-24-24zm64 236c0 6.6-5.4 12-12 12H108c-6.6 0-12-5.4-12-12v-8c0-6.6 5.4-12 12-12h168c6.6 0 12 5.4 12 12v8zm0-64c0 6.6-5.4 12-12 12H108c-6.6 0-12-5.4-12-12v-8c0-6.6 5.4-12 12-12h168c6.6 0 12 5.4 12 12v8zm0-72v8c0 6.6-5.4 12-12 12H108c-6.6 0-12-5.4-12-12v-8c0-6.6 5.4-12 12-12h168c6.6 0 12 5.4 12 12zm96-114.1v6.1H256V0h6.1c6.4 0 12.5 2.5 17 7l97.9 98c4.5 4.5 7 10.6 7 16.9z"
      />
    </svg>
    <span
      className="Filename"
    >
      test
    </span>
  </div>,
  <div
    className="fade alert alert-danger show"
    role="alert"
    style={
      Object {
        "marginTop": 10,
      }
    }
  />,
]
`;

exports[`renders Dropzone page 1`] = `
Array [
  <div
    className="Dropzone "
    onClick={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    <input
      accept=".csv"
      className="FileInput"
      id="uploadCaptureInputFile"
      onChange={[Function]}
      type="file"
    />
    <svg
      className="UploadIcon my-3"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 384 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M224 136V0H24C10.7 0 0 10.7 0 24v464c0 13.3 10.7 24 24 24h336c13.3 0 24-10.7 24-24V160H248c-13.2 0-24-10.8-24-24zm65.18 216.01H224v80c0 8.84-7.16 16-16 16h-32c-8.84 0-16-7.16-16-16v-80H94.82c-14.28 0-21.41-17.29-11.27-27.36l96.42-95.7c6.65-6.61 17.39-6.61 24.04 0l96.42 95.7c10.15 10.07 3.03 27.36-11.25 27.36zM377 105L279.1 7c-4.5-4.5-10.6-7-17-7H256v128h128v-6.1c0-6.3-2.5-12.4-7-16.9z"
      />
    </svg>
    <span
      className="Filename"
    >
      Choose a file
    </span>
  </div>,
  <div
    className="fade alert alert-danger show"
    role="alert"
    style={
      Object {
        "marginTop": 10,
      }
    }
  />,
]
`;
